cmd/agent/app/throttling/credit_store.go

```diff
@@ -0,0 +1,208 @@
+	perOperationBalance map[string]float64
+	defaultBalance      float64
+	updateTime          time.Time
+}
+
+func (c *client) Spend(operationName string, credits float64) bool {
+	balance, ok := c.perOperationBalance[operationName]
+	if !ok {
+		balance = c.defaultBalance
+	}
+
+	if credits > balance {
+		return false
+	}
+
+	balance -= credits
+	if ok {
+		c.perOperationBalance[operationName] = balance
+	} else {
+		c.defaultBalance = balance
+	}
+	return true
+}
+
+type CreditAccruerOptions struct {
+	MaxOperations    int
+	CreditsPerSecond float64
+	MaxBalance       float64
+	ClientMaxBalance float64
+}
+
+type creditAccruer struct {
+	options                 CreditAccruerOptions
+	serviceName             string
+	perOperationRateLimiter map[string]*tokenBucket
+	defaultRateLimiter      *tokenBucket
+	updateTime              time.Time
+	timeNow                 func() time.Time // For testing
+}
+
+func (ca *creditAccruer) Withdraw(operationName string, balance float64) float64 {
+	if balance >= ca.options.ClientMaxBalance {
+		return 0
+	}
+	rateLimiter, ok := ca.perOperationRateLimiter[operationName]
+	if !ok {
+		if len(ca.perOperationRateLimiter) < ca.options.MaxOperations {
+			rateLimiter = &tokenBucket{
+				creditsPerSecond: ca.options.CreditsPerSecond,
+				balance:          ca.options.MaxBalance,
+				maxBalance:       ca.options.MaxBalance,
+				timeNow:          ca.timeNow,
+				lastTick:         ca.timeNow(),
+			}
+			ca.perOperationRateLimiter[operationName] = rateLimiter
+		} else {
+			rateLimiter = ca.defaultRateLimiter
+		}
+	}
+
+	credits := rateLimiter.Drain(ca.options.ClientMaxBalance - balance)
+	return credits
+}
+
+type creditStore struct {
+	creditAccruers map[string]*creditAccruer
+	clients        map[string]*client
+	defaultOptions CreditAccruerOptions
+	lifetime       time.Duration
+	timeNow        func() time.Time // For testing
+}
+
+func NewCreditStore(defaultOptions CreditAccruerOptions, lifetime time.Duration) *creditStore {
+	s := &creditStore{
+		creditAccruers: map[string]*creditAccruer{},
+		clients:        map[string]*client{},
+		defaultOptions: defaultOptions,
+		lifetime:       lifetime,
+	}
+	s.timeNow = time.Now
+	return s
+}
+
+func (s *creditStore) Withdraw(clientID string, serviceName string, operationName string) float64 {
+	c := s.findOrCreateClient(clientID)
+	ca := s.findOrCreateCreditAccruer(serviceName)
+
+	// To keep client consistent with creditAccruer, we need to see if creditAccruer
+	// will create a new perOperationRateLimiter for this operation and if so
+	// create a new perOperationBalance on the client.
+	var balance float64
+	if _, ok := ca.perOperationRateLimiter[operationName]; ok {
+		balance = c.perOperationBalance[operationName]
+	} else {
+		if len(ca.perOperationRateLimiter) < ca.options.MaxOperations {
+			c.perOperationBalance[operationName] = balance
+		} else {
+			balance = c.defaultBalance
+		}
+	}
+
+	credits := ca.Withdraw(operationName, balance)
```

Comment:
maybe it's just me but I think this interface would be much simpler if rather than passing in the balance that the client already has, we pass in what it wants (ie ClientMaxBalance - balance) and if it's 0, we can just short circuit here. It makes creditAccruer more simpler (it shouldn't have to know about ca.options.ClientMaxBalance).

Reply:
Interesting idea. Will consider for next set of changes.